File: wallet/src/main/kotlin/org/trustnote/wallet/network/hubapi/HubResponse.java
package org.trustnote.wallet.network.hubapi;

import com.google.gson.JsonParser;


public class HubResponse extends HubPackageBase {

    public final static String TYPE_Justsaying = "justsaying";
    public final static String TYPE_Status_connected = "socketconnected";
    public final static String TYPE_Status_closed = "socketclosed";
    public final static String KEY_Subject = "mSubject";
    public final static String KEY_CHALLENGE = "hub/challenge";
    public final static String KEY_Body = "body";

    static HubResponse parseResponse(String hubMsg) {

        int index = hubMsg.indexOf(',');

        if (index < 3) {
            return null;
        } else {
            HubResponse res = new HubResponse();
            res.msgType = MSG_TYPE.valueOf(hubMsg.substring(2, index - 1));
            res.content = hubMsg.substring(index + 1, hubMsg.length() - 1);
            res.body = new JsonParser().parse(res.content).getAsJsonObject();
            if (TYPE_Justsaying.equals(res.msgType)) {
                if (KEY_CHALLENGE.equals(res.body.getAsJsonPrimitive(KEY_Subject).getAsString())) {
                    res.subjectType = BODY_TYPE.RES_CHALLENGE;
                }
            }
            return res;
        }
    }


    public static HubResponse createConnectedInstance() {
        HubResponse res = new HubResponse();
        res.msgType = MSG_TYPE.CONNECTED;
        return res;
    }

    public static HubResponse createCloseInstance() {
        HubResponse res = new HubResponse();
        res.msgType = MSG_TYPE.CLOSED;
        return res;
    }

    public String getChallenge() {
        if (subjectType != BODY_TYPE.RES_CHALLENGE) {
            throw new IllegalStateException("Msg is not RES_CHALLENGE");
        }
        return body.get(KEY_Body).getAsString();
    }
}

File: wallet/src/main/kotlin/org/trustnote/wallet/network/hubapi/HubResponse.kt
package org.trustnote.wallet.network.hubapi

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import org.trustnote.db.DbHelper
import org.trustnote.wallet.util.Utils
import org.trustnote.wallet.walletadmin.WalletModel


class HubResponse : HubMsg {

    val tag: String
    val responseJson: JsonElement

    constructor(textFromHub: String) : super(textFromHub) {
        tag = msgJson.getAsJsonPrimitive(HubMsgFactory.TAG).asString

        responseJson = msgJson.get(HubMsgFactory.RESPONSE) ?: Utils.emptyJsonObject
    }

    constructor(responseJson: JsonObject, tag: String) : super(MSG_TYPE.response) {
        this.msgSource = MSG_SOURCE.wallet
        this.responseJson = responseJson
        this.tag = tag
    }

    fun handResonse(hubSocketModel: HubSocketModel): Boolean {

        return hubSocketModel.responseArrived(this)
    }




}

